File: leetcode_problems/cpp/lc1475_final_price_with_special_price.cpp
#include <vector>
#include <stack>
#include <algorithm>
using std::stack;
using std::vector;

class Solution
{
public:
    vector<int> finalPrices(vector<int> &prices)
    {
        stack<int> st;
        vector<int> res(prices.size());
        for (int i = prices.size() - 1; i >= 0; i--)
        {
            int price = prices[i];
            while (!st.empty() && st.top() > price)
            {
                st.pop();
            }
            if (!st.empty())
            {
                res[i] = price - st.top();
            }
            else
            {
                res[i] = price;
            }
            st.push(price);
        }
        return res;
    }
};

File: leetcode_problems/typescript/src/lc1475_final_price_with_special_price.ts
function finalPrices(prices: number[]): number[] {
  let stack: number[] = [];
  let res = new Array(prices.length);
  for (let i = prices.length - 1; i >= 0; i--) {
    let price = prices[i];
    while (stack.length > 0 && stack[stack.length - 1] > price) {
      stack.pop();
    }
    if (stack.length > 0) {
      res[i] = price - stack[stack.length - 1];
    } else {
      res[i] = price;
    }
    stack.push(price);
  }
  return res;
}
